WordPress/gutenberg
# Full Site Editing: Add Support for Templates Default and Custom Titles and Descriptions (Second Attempt)

## Description

This is a simpler alternative to #26394, after an interesting [comment](https://github.com/WordPress/gutenberg/issues/26546#issuecomment-720183186) by @carlomanf made me realize that I was looking at this from the wrong side, and needlessly overcomplicating the approach.

Fixes #25755 

Update the templates handling to allow for custom titles and descriptions.

- Created a list of default template types definitions in PHP containing titles and descriptions for all the "generic" templates.
- The list is filterable with the new `default_template_types` filter, already used to temporarily remove the `embed` definition which it's currently not supported.
- The list is available in JS via Redux through the new `getDefaultTemplateTypes()`, `getDefaultTemplateType()`, and `getTemplateInfo()` selectors of `core/edit-site`.
- Templates imported from files will automatically obtain the title and excerpt from the default definitions, if possible.
- New templates created from the Site Editor sidebar will also automatically use the default definitions, and will be saved as `draft`. They will become `published` once saved the first time.
- The default definitions are used across the Site Editor as a fallback for templates without title or excerpt.
- Added support for `excerpts` to the `wp_template` CPT, used to store the template description.
- Updated the Templates wp-admin list to display more details (title, slug, excerpt, status), so that now it's at least slightly useful. It also now lists `auto-draft` templates, at least temporarily.
- Converted the `getTemplateInfo` utility into the `getTemplateInfo` selector, since it relies on the `core/edit-site` state.

### Main Differences with #26546

- No more `theme-provided` status. My original assumption that it was getting in the way of titles was proven incorrect after a deeper investigation. The status itself doesn't do anything to titles; it's the editor's entity retrieval that [clears the title](https://github.com/WordPress/gutenberg/blob/40a5502d62534cd64a623938754778501264736c/packages/core-data/src/actions.js#L85). In this PR there are some changes to prevent it.

- No more `theme` post meta to indicate the source of a template.
We shouldn't have duplicate templates, so a template can only ever be: either custom (`publish` or `draft` status), or provided by the **current** theme (`auto-draft` status).

### Follow up

- Update the template definitions copy!
- Add a way to edit the title (and maybe excerpt too) of a template from the Site Editor (e.g. in the Document Settings dropdown).
- Extend the wp-admin Templates list's quick edit with the excerpt (and maybe the slug too?).
- Update the template saving logic to allow for going back and forth between draft and publish.

## How has this been tested?

1. Open the Site Editor. Behind the scenes, all theme's and plugins' templates will be converted into new `wp_template` items of status `auto-draft`; there might be duplicates for unrelated reasons (see #26621, but not limited to it).
2. Open the wp-admin Templates list (`/edit.php?post_type=wp_template`).
3. Make sure it shows the following columns: title, slug, description, status.
4. Notice it shows also `auto-draft` templates. Feel free to wipe this list clean by moving all templates to trash, and permanently delete them.
5. Open the Site Editor (if you wiped the templates, now it will generate new ones, hopefully without duplicates).
6. Open the sidebar, and navigate into the Templates menu.
7. Make sure the items use the title and description as provided by the definitions list introduced in this PR.
8. Create a new template with the New Template dropdown (click on `+` in the Templates menu).
9. Make sure the items in the New Template dropdown use the definitions list.
10. Make sure the new template's name is the one provided by the definitions list (it should match the one from the New Template dropdown).
11. Check in the sidebar that the new template shows up (it might be easier to just look in the All Templates menu), it has the correct title and description, and is marked as "[Draft]".
12. Click on "Update Design" (it might be disabled: do some changes to enable it).
13. The template has now been published, so make sure it lost the "[Draft]" label.
14. Open the wp-admin Templates list, and make sure the new template shows up with the right title, description, and status.

## Screenshots 

The new Templates wp-admin list:

<img width="1117" alt="Screenshot 2020-11-02 at 15 35 37" src="https://user-images.githubusercontent.com/2070010/97895400-a2f9a100-1d2b-11eb-856c-140b5309d076.png">

The Site Editor sidebar using the new default definitions, and including a draft template:

<img width="522" alt="Screenshot 2020-10-26 at 17 44 59" src="https://user-images.githubusercontent.com/2070010/97208552-29a20180-17b3-11eb-892b-753f84693ffe.png">

The Site Editor documents settings using the new default definitions:

<img width="276" alt="Screenshot 2020-10-26 at 17 44 41" src="https://user-images.githubusercontent.com/2070010/97208602-3888b400-17b3-11eb-9848-9e6662943acc.png">

## Types of changes
Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
